client/order/epochqueue.go

```diff
@@ -0,0 +1,190 @@
+		seedIndex:       make([]order.OrderID, 0),
+		commitmentIndex: make([]order.OrderID, 0),
+	}
+}
+
+// Reset clears the epoch queue. This should be called when a new epoch begins.
+func (eq *EpochQueue) Reset() {
+	eq.ordersMtx.Lock()
+	eq.orders = make(map[order.OrderID]*msgjson.EpochOrderNote)
+	eq.ordersMtx.Unlock()
+
+	eq.seedIndexMtx.Lock()
+	eq.seedIndex = make([]order.OrderID, 0)
+	eq.seedIndexMtx.Unlock()
+
+	eq.commitmentIndexMtx.Lock()
+	eq.commitmentIndex = make([]order.OrderID, 0)
+	eq.commitmentIndexMtx.Unlock()
+}
+
+// Queue appends the provided note to the epoch queue.
+func (eq *EpochQueue) Queue(note *msgjson.EpochOrderNote) {
+	var oid order.OrderID
+	copy(oid[:], note.OrderID[:])
+
+	eq.ordersMtx.Lock()
+	defer eq.ordersMtx.Unlock()
+
+	eq.orders[oid] = note
+
+	// Update the seed sort order per order ids.
+	eq.seedIndexMtx.Lock()
+	si := sort.Search(len(eq.seedIndex), func(i int) bool {
+		order := eq.orders[eq.seedIndex[i]]
+		return bytes.Compare(order.OrderID[:], oid[:]) > 0
+	})
+	eq.seedIndex = append(eq.seedIndex, order.OrderID{})
+	copy(eq.seedIndex[si+1:], eq.seedIndex[si:])
+	eq.seedIndex[si] = oid
+	eq.seedIndexMtx.Unlock()
+
+	// Update the commitment sort order per order commitments.
+	eq.commitmentIndexMtx.Lock()
+	ci := sort.Search(len(eq.commitmentIndex), func(i int) bool {
+		order := eq.orders[eq.commitmentIndex[i]]
+		return bytes.Compare(order.Commitment[:], note.Commitment[:]) > 0
+	})
+	eq.commitmentIndex = append(eq.commitmentIndex, order.OrderID{})
+	copy(eq.commitmentIndex[ci+1:], eq.commitmentIndex[ci:])
+	eq.commitmentIndex[ci] = oid
+	eq.commitmentIndexMtx.Unlock()
+}
+
+// Exists checks if the provided order id is in the queue.
+func (eq *EpochQueue) Size() int {
+	eq.ordersMtx.Lock()
+	size := len(eq.orders)
+	eq.ordersMtx.Unlock()
+	return size
+}
+
+// Exists checks if the provided order id is in the queue.
+func (eq *EpochQueue) Exists(oid order.OrderID) bool {
+	eq.ordersMtx.Lock()
+	_, ok := eq.orders[oid]
+	eq.ordersMtx.Unlock()
+	return ok
+}
+
+// GenerateMatchProof calculates the sorting seed used in order matching as well as the
+// commitment checksum from the provided epoch queue preimages and misses.
+func (eq *EpochQueue) GenerateMatchProof(preimages []order.Preimage, misses []string) (int64, msgjson.Bytes, error) {
+	eq.ordersMtx.Lock()
+	defer eq.ordersMtx.Unlock()
+
+	// Remove all misses.
+	for _, entry := range misses {
+		b, err := hex.DecodeString(entry[:])
+		if err != nil {
+			return 0, nil, fmt.Errorf("decoding error: %v", err)
+		}
+		var oid order.OrderID
+		copy(oid[:], b[:])
+
+		delete(eq.orders, oid)
+
+		// Remove the seed index associated with the miss, preserving
+		// the sort order in the process.
+		eq.seedIndexMtx.Lock()
+		si := sort.Search(len(eq.seedIndex), func(i int) bool { return bytes.Compare(eq.seedIndex[i][:], oid[:]) >= 0 })
+		if si == len(eq.seedIndex) {
+			return 0, nil, fmt.Errorf("expected to find id %s in seed sort index", oid)
```

Comment:
If these errors happen, the entire EpochQueue is dead, correct?  The `eq.orders` slice has already had this oid removed, so you will get panics when attempting to get Commitment by looping through `eq.commitmentIndex` for `csum` computation in future calls to this function.  I suppose that's OK but it should be clearly documented that if `GenerateMatchProof` returns a non-nil error, the `EpochQueue` is hosed and `Reset` should be called, etc.

Reply:
Yes, for preimage mismatch errors and non existent orders for preimages the epoch queue should be reset.